homeassistant/components/camera/push.py

```diff
@@ -0,0 +1,181 @@
+                          config[CONF_BUFFER_SIZE],
+                          config[CONF_TIMEOUT])]
+
+    hass.http.register_view(CameraPushReceiver(cameras,
+                                               config[CONF_IMAGE_FIELD]))
+
+    async_add_devices(cameras)
+
+
+class CameraPushReceiver(HomeAssistantView):
+    """Handle pushes from remote camera."""
+
+    url = "/api/camera_push/{entity_id}"
+    name = 'api:camera_push:camera_entity'
+
+    def __init__(self, cameras, image_field):
+        """Initialize CameraPushReceiver with camera entity."""
+        self._cameras = cameras
+        self._image = image_field
+
+    async def post(self, request, entity_id):
+        """Accept the POST from Camera."""
+        try:
+            (_camera,) = [camera for camera in self._cameras
+                          if camera.entity_id == entity_id]
+        except ValueError:
+            _LOGGER.error("Unknown push camera %s", entity_id)
+            return self.json_message('Unknown Push Camera',
+                                     HTTP_BAD_REQUEST)
+
+        try:
+            data = await request.post()
+            _LOGGER.debug("Received Camera push: %s", data[self._image])
+            await _camera.update_image(data[self._image].file.read(),
+                                       data[self._image].filename)
+        except ValueError as value_error:
+            _LOGGER.error("Unknown value %s", value_error)
+            return self.json_message('Invalid POST', HTTP_BAD_REQUEST)
+        except KeyError as key_error:
+            _LOGGER.error('In your POST message %s', key_error)
+            return self.json_message('{} missing'.format(self._image),
+                                     HTTP_BAD_REQUEST)
+
+
+class PushCamera(Camera):
+    """The representation of a Push camera."""
+
+    def __init__(self, name, buffer_size, timeout):
+        """Initialize push camera component."""
+        super().__init__()
+        self._name = name
+        self._last_trip = None
+        self._filename = None
+        self._expired = None
+        self._state = STATE_IDLE
+        self._timeout = timeout
+        self.queue = deque([], buffer_size)
+
+        self.queue.append(self._blank_image())
+        self._current_image = self.queue[0]
+
+    @classmethod
+    def _blank_image(cls):
+        from PIL import Image
+        import io
+
+        image = Image.new('RGB', BLANK_IMAGE_SIZE)
+
+        imgbuf = io.BytesIO()
+        image.save(imgbuf, "JPEG")
+
+        return imgbuf.getvalue()
+
+    @property
+    def state(self):
+        """Current state of the camera."""
+        return self._state
+
+    async def update_image(self, image, filename):
+        """Update the camera image."""
+        if self._state == STATE_IDLE:
+            self._last_trip = dt_util.utcnow()
+            self.queue.clear()
+
+        self._state = STATE_RECORDING
+        self._filename = filename
+        self.queue.append(image)
```

Comment:
I feel like this platform is getting over complicated. It should just show the last pushed image.

We should also store that image on disk and not in memory.

Reply:
With que, you could see a stream if you push multible images into the entity. Maybe default 1 and you can overwrite this with a option?

Reply:
This is abusing a feature of our camera integration. The camera get_image function should show a realtime stream of what a camera is showing.

We don't support replays in our camera integration right now. So we shouldn't hack it in.

Reply:
I feel strongly against writing to disk. Most people run HA on an SD card, it really bad to write continuously to the SD.

I share @pvizeli comment (that was my intention in the implementation, therefore the current default size of deque = 1). 

Let me remind you that the most common use case for this camera is not realtime, and most of the time the user will be faced with the last image (very pure context).

Reply:
It could be resolved by allow customized save image path, then point it to some tmpfs.

Reply:
> Let me remind you that the most common use case for this camera is not realtime

I don't like this assumption. If you think you need a well-tuned components only for your use case, probably should not use a such generic name.

Reply:
@awarecan sure it can (thats how I did before developing this code)

But that's not easy for the average user, and would require hassio/hassos to be aware of that need.

We are not discussing full length video, we are discussing 3 or 5 images (it doesn't take that much memory)

Reply:
@awarecan I first had this component living in my custom_components repo (name is http_post). But realised people were using it for the described use case.

There is nothing specific in the component related to motionEye, so I did not call it camera.motionEye and even made some changes to make it even more general.

Reply:
A Raspberry Pi has 1GB of RAM. If 10 images are kept around, 1MB each, we'll be using 1% of memory for this feature.

Reply:
@balloob what is now the plan, remove or hold?